From 7e591cce95cbb06256e278c3617ea4cb7b01b6ef Mon Sep 17 00:00:00 2001
From: Zhu Ken <guigenz@codeaurora.org>
Date: Wed, 14 Jun 2017 11:45:28 -0700
Subject: [PATCH 223/500] bridge: fix eapol packet dropping issue

wpa_supplicant would receive EAPOL packet to start 802.1x session even if
its wireless STA interface enslaved in a bridge is disabled.

Change-Id: I4e3f786049a19c5bc7aacdc0528b0099744e8674
Signed-off-by: Zhu Ken <guigenz@codeaurora.org>
---
 net/bridge/br_input.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -451,13 +451,15 @@ forward:
 
 	switch (p->state) {
 	case BR_STATE_DISABLED:
-		if (ether_addr_equal(p->br->dev->dev_addr, dest))
-			skb->pkt_type = PACKET_HOST;
+		if (skb->protocol == htons(ETH_P_PAE)) {
+			if (ether_addr_equal(p->br->dev->dev_addr, dest))
+				skb->pkt_type = PACKET_HOST;
 
-		if (BR_HOOK(NFPROTO_BRIDGE, NF_BR_PRE_ROUTING,
-			dev_net(skb->dev), NULL, skb, skb->dev, NULL,
-			br_handle_local_finish) == 1) {
-			return RX_HANDLER_PASS;
+			if (BR_HOOK(NFPROTO_BRIDGE, NF_BR_PRE_ROUTING,
+				dev_net(skb->dev), NULL, skb, skb->dev, NULL,
+				br_handle_local_finish) == 1) {
+				return RX_HANDLER_PASS;
+			}
 		}
 		break;
 
